### PR TITLE
Add dynamic shape enabling operations: `ops.shape`, `ops.size`, and `ops.at` to the  `FusionDefinition`

### PR DESCRIPTION
**Pull request by kevinstephano**
This PR is adding the following operations to the FusionDefinition:
1. `fd.ops.shape(Tensor)`: returns a `Vector` of symbolic tensor dimension sizes.  Includes a `Tensor` method (ie `T1.shape()`).
2. `fd.ops.size(Tensor, dim)`: returns a symbolic size for a given `Tensor` dimension.  Includes a `Tensor` method (ie `T1.size(0)`).
3. `fd.ops.at(Vector,  index)`: returns a symbolic value for a given `Vector` index.  Includes a `Vector` method `__getitem__` that supports bracket indexing (ie `V1[0]`).

This PR also includes:

- Opinfo error testing of the 3 new functions.
- Arith functions in the code generator for each function

### CODE REVIEW

File: python_tests/pytest_opinfos.py [hunk 1]
```diff
@@ -108,6 +108,44 @@
...
+""" Start Dynamic Shape Enabling Operations """
+
+dynamic_shapes_ops = []
+
+# TODO: Add correctness testing as noted below
+vector_at_opinfo = OpInfo(
+    lambda fd: fd.ops.at,
+    "vector_at",
+    # TODO: Check correctness once there are operators that can consume a Vector
+    sample_input_generator=None,
```
Comment (rdspring1) [inline python_tests/pytest_opinfos.py:120]: Suggestion: You can return from an input generator a `vector` or `tensor` input and a `number` or `list` output.

**Comment by kevinstephano** [comment]
!build

File: csrc/ops/arith.cpp [hunk 2]
```diff
@@ -1570,6 +1570,50 @@ std::vector<Val*> tensor_sizes(TensorView* inp) {
   return sizes;
 }
 
+std::vector<Val*> shape(TensorView* inp) {
+  auto iter_domains = TensorDomain::noReductions(inp->getMaybeRFactorDomain());
+  std::vector<Val*> shape(iter_domains.size(), nullptr);
+
+  for (auto idx : c10::irange(iter_domains.size())) {
+    shape[idx] = iter_domains[idx]->getMaybeExpandedExtent();
+  }
```
Comment (jacobhinkle) [inline csrc/ops/arith.cpp:1579]: Style suggestion. Feel free to ignore:
```suggestion
  std::vector<Val*> shape;
  shape.reserve(iter_domains.size());
  for (auto id : iter_domains) {
    shape.push_back(id->getMaybeExpandedExtent());
  }
```

File: python_tests/pytest_input_generators.py [hunk 3]
```diff
@@ -13,6 +13,11 @@
 from pytest_utils import make_number, find_nonmatching_dtype, is_floating_dtype
 from nvfuser import DataType
 
+MINIMUM_SYMBOLIC_SIZE = -1
+INT64_MAX = 9223372036854775807
```
Comment (jacobhinkle) [inline python_tests/pytest_input_generators.py:17]: `2 ** 63 - 1` might be a little clearer.

File: python_tests/pytest_input_generators.py [hunk 4]
```diff
@@ -365,6 +360,15 @@ def define_vector_input_error_generator(
...
 
 
+def tensor_at_error_generator(
+    op: OpInfo, dtype: torch.dtype, requires_grad: bool = False, **kwargs
+):
+    """
+    [&at_def](Vector arg, int64_t index) -> Scalar {
+      return at_def(arg, index);
+    """
+
```
Comment (jacobhinkle) [inline python_tests/pytest_input_generators.py:370]: Should this be removed?

File: csrc/python_frontend/python_bindings.cpp [hunk 5]
```diff
@@ -2083,6 +2068,86 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+      [&at_def](Vector arg, int64_t index) -> Scalar {
+        return at_def(arg, index);
+      },
+      py::return_value_policy::reference);
+  vector_class.def(
+      "__getitem__",
+      [&at_def](Vector arg, int64_t index) -> Scalar {
+        return at_def(arg, index);
+      },
+      py::return_value_policy::reference);
```
Comment (jacobhinkle) [inline csrc/python_frontend/python_bindings.cpp:2142]: These dunder methods make it much easier to write fusion definitions. Another one you might consider adding is `__len__`. Note that even without `__len__` you can enable iteration over `Vector` in Python by raising `pybind11::index_error` (which is `IndexError` in Python) for `index >= size()`.

**Review by jacobhinkle — APPROVED** [review]
LGTM. Just had a few minor comments.

File: csrc/python_frontend/fusion_record.h [hunk 6]
```diff
@@ -2476,6 +2476,150 @@ struct TensorSizesRecord : RecordFunctor {
...
+//! Uses the default hash() and print() methods of Record Functor
+
+struct SizeOpRecord : RecordFunctor {
+  SizeOpRecord(std::vector<State> args, std::vector<State> outputs, int64_t dim)
+      : RecordFunctor(
+            std::move(args),
+            std::move(outputs),
+            "ops.size",
+            serde::RecordType_SizeOp),
+        dim_(dim) {}
```
Comment (jjsjann123) [inline csrc/python_frontend/fusion_record.h:2519]: should we check `dim_` here being within the range?

Comment (jjsjann123) [reply csrc/python_frontend/fusion_record.h:2519]: ^^^ Since this throws an error earlier rather than later when we construct the fusion. But this is only a nitpick.

**Review by jjsjann123 — APPROVED** [review]
minor comments. LGTM.

File: csrc/ops/arith.h [hunk 7]
```diff
@@ -416,6 +416,13 @@ TORCH_CUDA_CU_API TensorView* expand_as(TensorView* inp, TensorView* other);
 // with functions like broadcast_in_size that take in a vector of sizes
 // to use to expand an input tensor
 TORCH_CUDA_CU_API std::vector<Val*> tensor_sizes(TensorView* inp);
+// This is a function used to give the symbolic shape of a tensor for use
+// with functions like broadcast_in_size that take a shappe vector
```
Comment (rdspring1) [inline csrc/ops/arith.h:420]:
```suggestion
// with functions like broadcast_in_size that take a shape vector
```

**Review by zasdfgbnm — APPROVED** [review]

File: csrc/python_frontend/fusion_record.h (hunk 6 above)
Comment (kevinstephano) [reply csrc/python_frontend/fusion_record.h:2519]: I'll think about this recommendation.  I am not sure if I will make a change in this PR.

File: csrc/python_frontend/python_bindings.cpp (hunk 5 above)
Comment (kevinstephano) [reply csrc/python_frontend/python_bindings.cpp:2142]: I'll think about this recommendation.  I am not sure if I will make a change in this PR.

**Comment by kevinstephano** [comment]
!build
